### PR TITLE
refactor: use a struct instead of map[string]interface{}

### PR DESCRIPTION
**Pull request by luisdavim**
* Use a `struct` instead of `map[string]interface{}`
* Use the actual object instead of the last applied configuration annotation

~~If you think checking the `kubectl` annotation is important, I can add it as an extra check but I do think we should check the actual object first~~
**update:** I've added a new flag to toggle checking the `kubectl` annotation

Note: I had to move `Version` from the `judge` package to the `collector` package to avoid cyclic dependencies, maybe version should be its own package?

I've tested a build from my branch against a real EKS cluster with version 1.23:

```console
$ ./bin/kubent-darwin-amd64  
8:06PM INF >>> Kube No Trouble `kubent` <<<
8:06PM INF version dev (git sha dev)
8:06PM INF Initializing collectors and retrieving data
8:06PM INF Target K8s version is 1.23.10-eks-15b7512
8:06PM INF Retrieved 930 resources from collector name=Cluster
8:06PM INF Retrieved 0 resources from collector name="Helm v2"
8:07PM INF Retrieved 157 resources from collector name="Helm v3"
8:07PM INF Loaded ruleset name=custom.rego.tmpl
8:07PM INF Loaded ruleset name=deprecated-1-16.rego
8:07PM INF Loaded ruleset name=deprecated-1-22.rego
8:07PM INF Loaded ruleset name=deprecated-1-25.rego
8:07PM INF Loaded ruleset name=deprecated-future.rego
__________________________________________________________________________________________
>>> Deprecated APIs removed in 1.25 <<<
------------------------------------------------------------------------------------------
KIND                  NAMESPACE     NAME                                        API_VERSION      REPLACE_WITH (SINCE)
PodDisruptionBudget   kube-system   cluster-autoscaler-aws-cluster-autoscaler   policy/v1beta1   policy/v1 (1.21.0)
PodDisruptionBudget   kube-system   ingress-nginx-controller                    policy/v1beta1   policy/v1 (1.21.0)
PodDisruptionBudget   kube-system   ingress-nginx-internal-controller           policy/v1beta1   policy/v1 (1.21.0)
PodDisruptionBudget   kyverno       kyverno                                     policy/v1beta1   policy/v1 (1.21.0)
PodSecurityPolicy     <undefined>   eks.privileged                              policy/v1beta1   <removed> (1.21.0)
PodSecurityPolicy     <undefined>   pl                                          policy/v1beta1   <removed> (1.21.0)
```

### CODE REVIEW

**Comment by stepanstipl** [comment]
Hi @luisdavim, thanks for this 👍.

First, there are a bit too many things mixed in one PR. Due to that I won't review everything, but will comment on the main points:

> Use the actual object instead of the last applied configuration annotation

This is a bit of a fundamental misconception of how K8S API, and subsequently _kubent_, works. The version in the resource itself can't be used as it's governed by the client. I have explained this a few times in the tickets, please see [^1] for more details and further links. 

> update: I've added a new flag to toggle checking the kubectl annotation

You can disable the cluster collector, using the -`c=false` flag, but having a cluster collector without checking the last-applied annotation wouldn't make any sense.

> Use a struct instead of map[string]interface{}

I'm failing to see benefits of this change - i.e. we're introducing a custom struct, which is not equivalent. The `map[string]interface{}` would hold the whole object, while the new struct would just hold metadata.

Also at a quick first look, I've noticed some issues with the implementation, such as dropping support for parsing JSON files, or replacing the yaml library.

Having said that, and while I'm not a fan of the major changes above (or perhaps I don't understand the benefits in case of the struct/replacing the yaml library), some of the code cleanups look nice, and I would be happy to accept those. But the main point would be please open a separate PR for each type of change that you think is desireable, so that we can discuss and review it individually, without being entangled with all the other changes.

Cheers 👍 


[^1]: https://github.com/doitintl/kube-no-trouble/issues/357#issuecomment-1245338759

**Comment by luisdavim** [comment]
Hi, thanks for the feedback, some notes:

* I didn't drop support for `JSON` I've switched to the k8s `YAML` lib that handles both by first converting yaml to json and then do the marsheling/unmarsheling
* It's true that by using `map[string]interface{}` you get the full object, but that's not being used so using a struct to capture what we actually need from the object makes the code cleaner and simpler
* The flag is to disable the use of the `kubectl` annotation and use the actual object
* A big advantage of this change is also that allowed displaying the collector name, so users don't get confused when the tool complains about some resource that has been updated or doesn't even exist like in #310 I know about how objects are stored in `etcd` but if they can be retrieved with the supported version they are not a problem, that means they were converted, this is why it's important to check both and print the source (collector) so the user can identify false positives, having an object that is using a supported API but has the annotation with the unsupported API tells the users they should check the source file but if they don't know the source they just get confused.
* I can break this into separate PRs if you are still interested in this.

Here's an example output from my latest commit:

```console
$ go run cmd/kubent/main.go                                    
4:57PM INF >>> Kube No Trouble `kubent` <<<
4:57PM INF version dev (git sha dev)
4:57PM INF Initializing collectors and retrieving data
4:57PM INF Target K8s version is 1.23.10-eks-15b7512
4:57PM INF Retrieved 1022 resources from collector name=Cluster
4:57PM INF Retrieved 0 resources from collector name="Helm v2"
4:57PM INF Retrieved 157 resources from collector name="Helm v3"
4:57PM INF Loaded ruleset name=custom.rego.tmpl
4:57PM INF Loaded ruleset name=deprecated-1-16.rego
4:57PM INF Loaded ruleset name=deprecated-1-22.rego
4:57PM INF Loaded ruleset name=deprecated-1-25.rego
4:57PM INF Loaded ruleset name=deprecated-future.rego
__________________________________________________________________________________________
>>> Deprecated APIs removed in 1.25 <<<
------------------------------------------------------------------------------------------
KIND                  NAMESPACE     NAME                                        API_VERSION      REPLACE_WITH (SINCE)   COLLECTOR
PodDisruptionBudget   kube-system   cluster-autoscaler-aws-cluster-autoscaler   policy/v1beta1   policy/v1 (1.21.0)     Helm v3
PodDisruptionBudget   kube-system   ingress-nginx-controller                    policy/v1beta1   policy/v1 (1.21.0)     Helm v3
PodDisruptionBudget   kube-system   ingress-nginx-internal-controller           policy/v1beta1   policy/v1 (1.21.0)     Helm v3
PodDisruptionBudget   kyverno       kyverno                                     policy/v1beta1   policy/v1 (1.21.0)     Helm v3
PodSecurityPolicy     <undefined>   eks.privileged                              policy/v1beta1   <removed> (1.21.0)     Cluster-annotation
PodSecurityPolicy     <undefined>   eks.privileged                              policy/v1beta1   <removed> (1.21.0)     <undefined>
PodSecurityPolicy     <undefined>   pl                                          policy/v1beta1   <removed> (1.21.0)     Cluster
```

**Comment by stepanstipl** [comment]
Hi @luisdavim,

thanks for getting back and clarifying some of the points. I will comment only on the core problem:

> The flag is to disable the use of the kubectl annotation and use the actual object
> 
> A big advantage of this change is also that allowed displaying the collector name, so users don't get confused when the tool complains about some resource that has been updated or doesn't even exist like in https://github.com/doitintl/kube-no-trouble/issues/310 I know about how objects are stored in etcd but if they can be retrieved with the supported version they are not a problem, that means they were converted, this is why it's important to check both and print the source (collector) so the user can identify false positives, having an object that is using a supported API but has the annotation with the unsupported API tells the users they should check the source file but if they don't know the source they just get confused.

This is a misconception of how K8S versioning works, and this change doesn't make sense. As I've explained[^1], and more in detail[^2], the returned version is governed by the client and not stored anywhere. I.e. you can get the same object in two versions:

> ...
> $ kubectl get ingress.v1beta1.networking.k8s.io test -o yaml
> Warning: networking.k8s.io/v1beta1 Ingress is deprecated in v1.19+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
> apiVersion: networking.k8s.io/v1beta1
> kind: Ingress
> metadata:
> ...
> 
> and also
> 
> $ kubectl get ingress.v1.networking.k8s.io test -o yaml
> apiVersion: networking.k8s.io/v1
> kind: Ingress
> metadata:
>  ...

I.e. should we only look at the `apiVersion` of the returned object, it will be governed by the API version we request[^3], and never get anything deprecated. I.e. the cluster collector would never return anything and therefore be a bit pointless.

[^1]: https://github.com/doitintl/kube-no-trouble/issues/276#issuecomment-1062082004
[^2]: https://github.com/doitintl/kube-no-trouble/issues/276#issuecomment-1062928787
[^3]: https://github.com/doitintl/kube-no-trouble/blob/master/pkg/collector/cluster.go#L73-L107

**Comment by luisdavim** [comment]
@stepanstipl I don't think you understood what I meant. 
The api version conversion is not done on the client-side. See https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-groups-and-versioning

> API resources are distinguished by their API group, resource type, namespace (for namespaced resources), and name. The API server handles the conversion between API versions transparently: all the different versions are actually representations of the same persisted data. The API server may serve the same underlying data through multiple API versions.

Users use this tool to check if they have resources in the cluster that could be a problem, not all the resources will be applied by `kubectl` or helm, some might be generated by a controller. Checking both the actual object and the annotations and reporting where the violation comes from (actual object, helm release, annotation, etc..) allows the user to understand if the issue is a false positive as well has the source, foe example, you can fix the latest helm release but know you can't rollback because the resources in a previous release are not supported, or maybe a resource was patched to fix the issue but the kubectl annotation was not updated, or the resource is managed by some other tool but at some point was edited with kubectl so it got an annotation that may now be stale.

What I'm doing here is to check both and let the user know which collector reported the issue, if the issue comes from an annotation instead of the actual object the source will be listed as `Cluster-annotation`

For example, the issue that got me here was a resource that was managed by EKS that for some reason had a stale kubectl annotation, this change would still warn about it but make the debugging easier, I had to read the tool's code to understand why this was being reported as an issue.

**Comment by stepanstipl** [comment]
Hi @luisdavim please check the above example with Ingress. The *same* object will always be returned in the version requested by the client. I.e. same object -> 2 different API versions.

> The API server may serve the same underlying data through multiple API versions.

The API version of the returned object will be governed by our client and checking against those has no meaning as the deprecated versions would only be returned if we would have requested those in the first place (and we do not [^1]).

This is better described in [^2], but there is ever only one internal version of a given resource. However, resources can be created in any of the supported API versions and conversion to internal format will happen. And the other way when retrieving the resource. But again, the returned API version is governed by _kubent_ as a client, i.e. we get what we request and checking against that makes no sense.

I understand this can be confusing for users, but unfortunately, that's how K8S API works in the first place. _Some of the usability improvements such as displaying the collector column are potentially nice, and we can discuss those separately._

 _If it would be as simple as checking the returned `apiVersion`, this tool would not exist in the first place. 😉_

[^1]: https://github.com/doitintl/kube-no-trouble/blob/master/pkg/collector/cluster.go#L73-L107 [↩](https://github.com/doitintl/kube-no-trouble/pull/370#user-content-fnref-3-59686eb3c7ba9a0c57b947fd29e94c65)
[^2]: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md

**Comment by luisdavim** [comment]
Yes @stepanstipl , I get that. As I've said, I understand how the resources are being stored in `etcd` and retreived fom the API but there are edge cases that the tool doesn't cover.

 If the API can convert the resrouce into the new version it won't be a problem and will not be reported unless the `kubectl` annotation has the deprecated API version. What I'm trying to cover with these changes are cases like the following:
* The resouce doesn't have a `kubectl` annotation and it's not part of any helm release and the API can't convert it, for example if you have some controller that generates PSPs (a resource that will be removed), without these changes that issue will not be reported.
* The resource can be converted and it's not really a problem but the `kubectl` annotation is referring to a deprecated API, in this case the tool shhould (and still wiil with these changes) report the issue but also let the user know it comes from the annotation and not the resource itself.

The point is, I think these changes make the tool more roubust and its output more informative. I'm not removing any functionality or behaviour, only making the tool also check the actual object and letting the user know which collector reported the issue so a better informed decision can be made.

**Comment by luisdavim** [comment]
Also note that I'm removing more code than I'm adding whilst sill adding and not removing functionality...

**Comment by stepanstipl** [comment]
> * Use the actual object ~~instead of~~ and the last applied configuration annotation
> ~~If you think checking the `kubectl` annotation is important, I can add it as an extra check but I do think we should check the actual object first~~

So this PR started as a suggestion to replace the annotation, with checking the returned API version directly, which as explained, does not work.

> The resouce doesn't have a kubectl annotation and it's not part of any helm release and the API can't convert it, for example if you have some controller that generates PSPs (a resource that will be removed), without these changes that issue will not be reported.

There's only one case - resource - currently where this would work (PSP). Conceptually I'm against implementing behaviour that introduces special handling for individual resources. It's more confusing than beneficial IMHO.

> The resource can be converted and it's not really a problem but the kubectl annotation is referring to a deprecated API, in this case the tool shhould (and still wiil with these changes) report the issue but also let the user know it comes from the annotation and not the resource itself.

Improving output is potentially a good feature. Atm. I'm not convinced this should be done by changing the default one, as the collector is usually not something the user should be concerned about, but perhaps the output can be made more configurable (e.g. specify what columns should be printed). It would be for a discussion on how to expose this to the user - but please via separate issue/PR. Discussing all this together is difficult.

> Also note that I'm removing more code than I'm adding whilst sill adding and not removing functionality...

_This is mainly caused by removed tests, which are not adequately replaced (e.g. no tests for `parseManifest()`). 😉 And there are a few other implementation issues._

To sum up - the core change of this PR is IMHO misconception of how K8S API and kubent work and not desirable. I would be happy to discuss other proposed changes (already commented on most of those), but please let's do this via small targeted Issues/PRs so that it's more feasible to have a focused conversation about and review each of those. I'm not going to respond to this PR anymore to protect my time. Thanks.

**Comment by luisdavim** [comment]
This would bot be to cover the PSP specifically but to cover the case of removed APIs.
